Add Pod Security Standards documentation (#28903)
kubernetes/website: content/en/docs/setup/best-practices/enforcing-pod-security-standards.md

---
reviewers:
- tallclair
title: Enforcing Pod Security Standards
weight: 40
---

<!-- overview -->

This page provides an overview of best practices when it comes to enforcing [Pod Security Standards](/docs/concepts/security/pod-security-standards).

<!-- body -->

> ## Configure all cluster namespaces

Reviewer: These best practices assume that you're using a mechanism based on namespace-level restrictions, but don't mention the built-in mechanism.

I think we should let readers know:
- there are 3rd party alternatives
- the PodSecurity admission mechanism is in alpha
- they could also use the (deprecated) PodSecurityPolicy feature

This page should help readers pick which approach to adopt given their circumstances. It's a tricky one because an alpha feature might not feel right for everyone, yet. We can explain that, provided readers are happy to handle the potential changes from minor release to minor release, the new admission controller is the best in-project mechanism.

Is there any generic advice about how to use or adopt 3rd party tools (OPA / Gatekeeper etc)? It might be as simple as “make sure you do _something_” or “make sure you trust your supply chain”.

In an ideal world, every pod in every namespace would meet the requirements of the `restricted` policy. However, this is not possible nor practical, as some workloads will require elevated privileges for legitimate reasons.

- Namespaces allowing `privileged` workloads should establish and enforce appropriate access controls.
- For workloads running in those permissive namespaces, maintain documentation about their unique security requirements. If at all possible, consider how those requirements could be further constrained.

## Adopt a multi-mode strategy

The `audit` and `warn` modes of the Pod Security Standards admission controller make it easy to collect important security insights about your pods without breaking existing workloads.

It is good practice to enable these modes for all namespaces, except maybe for those that will _absolutely_ always require `privileged` access or those that enforce the `restricted` policy. Further, these modes should usually be set to policies at least one level higher than the policy for the `enforce` mode.

- Namespaces enforcing the `privileged` policy (i.e. no restrictions) should set audit and warn labels to `baseline`.

> - Namespaces enforcing the `baseline` policy should set audit and warn labels to `restricted`.

Reviewer 2: I don't agree these should always be set, or should usually be one level higher... that will create a lot of audit/warning noise if done indiscriminately.

I would say that audit and warn should be set to the desired enforce level/version when the enforced level/version is not yet to that level.

Once enforce is at the desired level/version, there are a few ways to use audit/warn mode:
* Setting warn mode _matching_ enforce mode surfaces warnings to users when creating workload objects whose pod templates don't fit inside the enforced level/version
* In a namespace that pins enforce to a specific non-latest version, setting audit and warn mode to the same level as enforce but to the `latest` version gives visibility to things that were allowed by previous versions but are not allowed by current best practices.

---
reviewers:
- tallclair
- liggitt
title: Enforcing Pod Security Standards
weight: 40
---

<!-- overview -->

This page provides an overview of best practices when it comes to enforcing [Pod Security Standards](/docs/concepts/security/pod-security-standards).

<!-- body -->

## Using the built-in Pod Security Admission Controller

{{< feature-state for_k8s_version="v1.22" state="alpha" >}}

The [Pod Security Admission Controller](/docs/reference/access-authn-authz/admission-controllers/#podsecurity) intends to replace the deprecated PodSecurityPolicies. 

### Configure all cluster namespaces

> Even if the initial configurations are most permissive (i.e. everything is `privileged`), any Namespaces that lack any configuration at all will become significant gaps in your cluster security model.

Reviewer 2: I'm not sure I agree with labeling all namespaces as `privileged` as a starting point... if I were migrating a cluster to this, I'd use unlabeled namespaces to represent "not yet evaluated", distinct from "evaluated and explicitly granted `privileged`"

We provide an [example](/docs/concepts/security/pod-security-admission/#applying-to-all-namespaces) that illustrates how you can do this.

### Embrace the principle of least privilege

In an ideal world, every pod in every namespace would meet the requirements of the `restricted` policy. However, this is not possible nor practical, as some workloads will require elevated privileges for legitimate reasons.

- Namespaces allowing `privileged` workloads should establish and enforce appropriate access controls.
- For workloads running in those permissive namespaces, maintain documentation about their unique security requirements. If at all possible, consider how those requirements could be further constrained.

### Adopt a multi-mode strategy

The `audit` and `warn` modes of the Pod Security Standards admission controller make it easy to collect important security insights about your pods without breaking existing workloads.

> It is good practice to enable these modes for all namespaces, setting them to the _desired_ level and version you would eventually like to `enforce`. The warnings and annotations generated in this phase can guide you toward that state.

Reviewer 2: ```suggestion
It is good practice to enable these modes for all namespaces, setting them to the _desired_ level and version you would eventually like to `enforce`. The warnings and audit annotations generated in this phase can guide you toward that state. If you expect workload authors to make changes to fit within the desired level, enable the `warn` mode. If you expect to use audit logs to monitor/drive changes to fit within the desired level, enable the `audit` mode. 
```

When you have the `enforce` mode set to your desired value, these modes can still be useful in a few different ways:

- By setting `warn` to the same level as `enforce`, clients will receive warnings when attempting to create Pods (or resources that have Pod templates) that do not pass validation. This will help them update those resources to become compliant.
- In Namespaces that pin `enforce` to a specific non-latest version, setting the `audit` and `warn` modes to the same level as `enforce`, but to the `latest` version, gives visibility into settings that were allowed by previous versions but are not allowed per current best practices.

## Third-party alternatives

{{% thirdparty-content %}}

Other alternatives for enforcing security profiles are being developed in the Kubernetes
ecosystem:

- [OPA Gatekeeper](https://github.com/open-profile-agent/gatekeeper).

> The decision to go with a _native_ solution (e.g. PodSecurity admission controller) versus a

Reviewer 2: ```suggestion
The decision to go with a _built-in_ solution (e.g. PodSecurity admission controller) versus a
```

---
reviewers:
- tallclair
- liggitt
title: Enforcing Pod Security Standards
weight: 40
---

<!-- overview -->

This page provides an overview of best practices when it comes to enforcing [Pod Security Standards](/docs/concepts/security/pod-security-standards).

<!-- body -->

## Using the built-in Pod Security Admission Controller

{{< feature-state for_k8s_version="v1.22" state="alpha" >}}

> The Pod Security Admission Controller intends to replace the deprecated PodSecurityPolicies. 

Reviewer: I'd link to `/docs/reference/access-authn-authz/admission-controllers/#podsecurity` here.

### Adopt a multi-mode strategy

The `audit` and `warn` modes of the Pod Security Standards admission controller make it easy to collect important security insights about your pods without breaking existing workloads.

It is good practice to enable these modes for all namespaces, setting them to the _desired_ level and version you would eventually like to `enforce`. The warnings and annotations generated in this phase can guide you toward that state.

When you have the `enforce` mode set to your desired value, these modes can still be useful in a few different ways:

- By setting `warn` to the same level as `enforce`, clients will receive warnings when attempting to create Pods (or resources that have Pod templates) that do not pass validation. This will help them update those resources to become compliant.
- In Namespaces that pin `enforce` to a specific non-latest version, setting the `audit` and `warn` modes to the same level as `enforce`, but to the `latest` version, gives visibility into settings that were allowed by previous versions but are not allowed per current best practices.

> ## Using PodSecurityPolicies

Reviewer: We can omit this heading and add more docs post-release.

## Third-party alternatives

> 

Reviewer: ```suggestion

{{% thirdparty-content %}}

```
(adds a canned disclaimer)

and version you would eventually like to `enforce`. The warnings and audit annotations generated in
this phase can guide you toward that state. If you expect workload authors to make changes to fit
within the desired level, enable the `warn` mode. If you expect to use audit logs to monitor/drive
changes to fit within the desired level, enable the `audit` mode.

When you have the `enforce` mode set to your desired value, these modes can still be useful in a
few different ways:

- By setting `warn` to the same level as `enforce`, clients will receive warnings when attempting
  to create Pods (or resources that have Pod templates) that do not pass validation. This will help
  them update those resources to become compliant.
- In Namespaces that pin `enforce` to a specific non-latest version, setting the `audit` and `warn`
  modes to the same level as `enforce`, but to the `latest` version, gives visibility into settings
  that were allowed by previous versions but are not allowed per current best practices.

## Third-party alternatives

{{% thirdparty-content %}}

Other alternatives for enforcing security profiles are being developed in the Kubernetes
ecosystem:

> - [OPA Gatekeeper](https://github.com/open-profile-agent/gatekeeper).

Reviewer 3: ```suggestion
- [OPA Gatekeeper](https://github.com/open-policy-agent/gatekeeper).
```

Author: Good catch! Fixed.
